src/App/Fossa/VSI/Analyze.hs

```diff
@@ -0,0 +1,352 @@
+import Data.Foldable (traverse_)
+import Data.Map qualified as Map
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Discovery.Archive (withArchive')
+import Discovery.Filters (AllFilters, combinedPaths, excludeFilters, includeFilters)
+import Discovery.Walk (WalkStep (WalkContinue, WalkSkipAll), walk)
+import Effect.Logger (Color (..), Logger, Severity (SevError), annotate, color, logDebug, logInfo, pretty)
+import Effect.ReadFS (ReadFS)
+import Fossa.API.Types (ApiOpts)
+import Path (Abs, Dir, File, Path, Rel, SomeBase (Abs, Rel), isProperPrefixOf, toFilePath, (</>))
+import Path qualified as P
+import Path.Extra (tryMakeRelative)
+
+runVsiAnalysis ::
+  ( Has (Lift IO) sig m
+  , Has Finally sig m
+  , Has Logger sig m
+  , Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has StickyLogger sig m
+  ) =>
+  Path Abs Dir ->
+  ApiOpts ->
+  ProjectRevision ->
+  AllFilters ->
+  m ([VSI.Locator], [IAT.UserDep])
+runVsiAnalysis dir apiOpts projectRevision filters = context "VSI" $ do
+  capabilities <- sendIO getNumCapabilities
+
+  scanID <- context "Create scan in backend" $ vsiCreateScan apiOpts projectRevision
+  logInfo . pretty $ "Created Scan ID: " <> unScanID scanID
+
+  -- Allow up to 2x the buffer size to be stored:
+  -- - runBatchUploader buffers up to uploadBufferSize
+  -- - newTBMChanIO also buffers up to uploadBufferSize
+  files <- sendIO $ newTBMChanIO uploadBufferSize
+  context "Fingerprint files"
+    . withTaskPool capabilities (updateProgress "Upload remaining files")
+    . runAtomicCounter
+    $ do
+      context "Discover fingerprints" . forkTask $ runFingerprintDiscovery capabilities files dir filters
+      context "Upload fingerprints" . forkTask $ runBatchUploader files apiOpts scanID
+
+  logInfo "Finalizing scan"
+  context "Finalize scan" $ vsiCompleteScan apiOpts scanID
+
+  logInfo "Waiting for cloud analysis"
+  context "Wait for cloud analysis" $ waitForAnalysis apiOpts scanID
+
+  discoveredRawLocators <- context "Download analysis results" $ vsiDownloadInferences apiOpts scanID
+  parsedLocators <- context "Parse analysis results" . fromEither $ traverse VSI.parseLocator discoveredRawLocators
+
+  let userDefinedDeps = map IAT.toUserDep $ filter VSI.isUserDefined parsedLocators
+  let allOtherDeps = filter (not . VSI.isUserDefined) parsedLocators
+  pure (allOtherDeps, userDefinedDeps)
+
+uploadBufferSize :: Int
+uploadBufferSize = 1000
+
+runFingerprintDiscovery ::
+  ( Has (Lift IO) sig m
+  , Has Finally sig m
+  , Has Logger sig m
+  , Has ReadFS sig m
+  , Has StickyLogger sig m
+  ) =>
+  -- | Parallelization factor
+  Int ->
+  -- | Channel on which new files are collected
+  TBMChan (Path Rel File, Combined) ->
+  Path Abs Dir ->
+  AllFilters ->
+  m ()
+runFingerprintDiscovery capabilities files dir filters = do
+  withTaskPool capabilities (updateProgress "Fingerprint files") . runAtomicCounter $ do
```

Comment:
[optional] if we "create" a new sticky logger here:

```hs
  runStickyLogger SevInfo . runTaskPool [...]
```

..rather than sending the constraint upward, we can eliminate the contention on the sticky logger output between this `updateProgress` and the `updateProgress` of the parent taskpool.

this will result in two separate "sticky" progress lines at the bottom of the terminal